src/NuGet.Clients/NuGet.SolutionRestoreManager/SolutionRestoreWorker.cs

```diff
@@ -251,6 +254,18 @@ public bool Restore(SolutionRestoreRequest request)
                         // Initialize if not already done.
                         await InitializeAsync();
 
+#if VS15
+                        if (request.RestoreSource == RestoreOperationSource.OnBuild)
+                        {
+                            if (!await SolutionManager.SolutionHasDeferredProjectsAsync()
+                                && SolutionManager.GetNuGetProjects().All(p => p is CpsPackageReferenceProject))
+                            {
```

Comment:
Why don't we move this check to `RestoreManagerPackage` itself and skip getting into SolutionRestoreWorker if it isn't requried at all, which will also save on unnecessary `JTF.Run` and `JoinableCollection.Join()`